### ISSUE TITLE
[Request] NIN - Consolidating Ninjutsus

### ISSUE THREAD

**Issue by baelrayneVessdren** [MKhayle/XIVComboExpanded#476]
Hi @MKhayle @Kaedys

I was wondering if something like this would be possible (left Ninjutsu out of the diagram)?

![image](https://github.com/user-attachments/assets/8d290fb4-2fdd-4502-90cb-ebc3e485373d)

IMO it would go well with this:
![image](https://github.com/user-attachments/assets/916a84c1-7c14-4609-a94a-617c308fbb14)

If I am not mistaken / too sleep deprived there should not be any overlap and since all Ninjutsus start with different Mudras, there should also be no issue with Kassatsu (I think).

As always, many thanks for your hard work. 

Best regards

**Comment by kaedys** [comment]
I took a look at this a few months back, and the central problem is, the Ninja gauge in Dalamud provides no way to track if you're in a Ninjutsu sequence, which of the mudras you've used, what your last one was, etc.  I can't really think of a clean way of handling it within the scope of this mod.  Beyond that, there's not really a clean way of indicating which of them you're trying to case, since you cannot add the end abilities to your bars, and even if you do so via a mod (ex. Reaction), they don't _work_ with the mod, as far as I can tell.

Also, as a point of note, the _order_ of the mudras doesn't make a lick of difference.  The _only_ things that affect which one you're using are the number of symbols used, and which one you _end_ on.  For example, your diagram lists Jin -> Ten as Raton, but Chi -> Ten will also cast Raton.  Similarly, both Jin > Ten > Chi and Ten > Jin > Chi will cast Dotan.

**Comment by baelrayneVessdren** [comment]
I know the last part. Which is why I have formatted only the last Mudra in Bold. Also to make more visible that there are no overlaps.

That there is no way to track the actual Mudra used is a pity. Could it be possible to hard code the sequence though? Also one of the reasons that the diagram looks the way it does. If the ability itself is hard coded in a sequence (as any other 1-2-3 rotation I assume), the 2 Mudra Ninjutsu would also be always the same and you can use Ninjutsu as a a separate ability for those. That was the reason I also posted the Aeolian Edge / Armor Crush options. 

The Ninja gauge should not play into this IMO. Since you can't interrupt a Jutsu sequence if I am not mistaken. Any Weaponskill used should interrupt it.

**Comment by kaedys** [comment]
The way the Aeolian Edge and Armor Crush implementations work is if you have the Mudra buff (which is applied by the first symbol used, but does not refresh or stack), it swaps to Ninjutsu.  We can use that to detect if you're in a Mudra sequence, in theory (though buffs can be a bit laggy), but we can't determine which symbols you've used, or what your last one was.  I brought up the gauge because the Dalamud gauge implementation is how we track _most_ job-specific stuff like that.  Note that the _Dalamud_ gauge and the _in-game gauge_ are two very different things.  For example, the Dalamud gauge for Viper has the ability to track whether you're in a Vicewinder combo, and if so, which one.  It also has a field for tracking which Serpent's Tail ability is available.  Neither of those are on the in-game gauge, but the Dalamud implement tracks it.  I _assume_ the game tracks the Ninjutsu sequence for Ninjas in a similar manner as it does for Vicewinder combos, as part of the job combat data, but Dalamud has not implemented the fields to read that data from the client, so we don't really have a way to identify where in the sequence you're at.

See, the thing is, this mod doesn't have _sequences_.  Every button press, the combos determine which action to use next based on the data at the time of the button press.  What was the last combo action, what are your buff and resource and gauge states, etc.  The sequences that exist occur organically because of these conditions.  But there's no condition for us to detect, for example, that you just used Chi and are trying to cast Huton, so we need to use Jin.  We don't know you just used Chi, because it's not a "combo" action, and we also don't have a way for the player to indicate they want to use Huton specifically, of the 6 options available (again, excluding Shuriken).

**Comment by baelrayneVessdren** [comment]
That is interesting stuff. Thanks a lot for the thorough explanation. Not that versed in coding (don't have the brain for it), but I think I get the overall picture from your descriptions.

Can't be helped then. Have to memorize the possible combinations. My ADHD brain thanks all parties involved for the opportunity 🤣

**Comment by kaedys** [comment]
That was where I ended up, too.  Just have to memorize the combos😭.  I'm objectively _terrible_ at Ninja, and especially Ninjutsus, which was why I was trying to find a way to do this in the first place (when I was leveling NIN to 100 a couple months ago).  I ended up making an on-screen cheatsheet of the combos using DelvCD, which at least _helped_.  I can share the preset for it, if you're using the mod and would like it.

**Comment by baelrayneVessdren** [comment]
Same 😂 Overall NIN is not that hard imo. I believe that SAM is more "complicated" in terms of execution, if it weren't for those damn Jutsus. They ruin it for me 😂 

I'd appreciate the cheat sheet. Not using Delv as much as I used to, but that does not stop me from trying it out. Thanks a lot for that.

**Comment by kaedys** [comment]
Note that DelvCD and DelvUI are separate mods.  The former is more akin to the role Weakauras or TellMeWhen filled in WoW, if you're familiar with those.  Basically, the ability to display icons, bars, indicators, etc when certain conditions are fulfills, such as an ability being off cooldown or a debuff being present or a bunch of others like that.

I'll get the preset exported tomorrow and post it here.

**Comment by MKhayle** [comment]
so uhhhh uhhhhh should I close the issue or?
tl;dr is to use delvcd/ui?
